### PR TITLE
Allow locking documents copied to workspace

### PR DESCRIPTION
**Pull request by njohner**
In this PR we add the option of locking documents that get copied to a Workspace. Locking happens on the Gever side, to avoid modifying the document in Gever, as the intent is to create a new version of that document (and unlock it) when copying back the document from the workspace.

The whole mechanism and its implementation is very similar to the locking of documents submitted with a proposal, where we lock the documents on the meeting side, so that new versions can be uploaded from the meeting dossier.

Here we add the option to lock the document on the Gever side when copying it to a workspace. Lock information is displayed as for other locked documents and any user with permissions to modify the document can override the lock. Overriding the lock is done over a confirmation form.

We also adapt the API endpoint to support the additional parameter for locking the document when copying it to a workspace. This is a non-mandatory parameter and defaults to `False`, i.e. not locking the document.

Finally we realised that locked documents were not well tested and that there were several issues with locked documents, notably that a locked document could be trashed and that the move action was available on locked documents (although the action would fail). We therefore:
- Add tests for locked documents
- Forbid trashing locked documents
- Disable move action for locked documents

For https://4teamwork.atlassian.net/browse/CA-1221

**Displaying the locking info**
Screenshot 2020-11-18 at 08 59 56

**Unlock confirmation form**
Screenshot 2020-11-18 at 12 30 02

Definition of Done: https://4teamwork.atlassian.net/wiki/spaces/CHX4TW/pages/917562/

## Checklist (Must have)

_Everything has to be done/checked. Checked but not present means the author deemed it unnecessary._

- [x] Changelog entry
- [x] Link to issue (Jira or GitHub) and backlink in issue (Jira)

## Checklist (if applicable)

_Only applicable should be left and checked._

- API change:
  - [x] Documentation is updated
  - If breaking: **Not needed, as default is the same as before**
    - [ ] api-change label added: 
    - [ ] #delivery channel notified about breaking change
    - [ ] Scrum master is informed
- New functionality:
  - [ ] for `document` also works for `mail`
- New translations
  - [x] All msg-strings are unicode

### CODE REVIEW

**Review by lukasgraf — CHANGES_REQUESTED** [review]

**Comment by lukasgraf** [comment]
As discussed: Looks good overall, very clean and also easy to review! 👍  🎉 
Just those minor points regarding the exact messages, and two others:
- There don't seem to be tests that verify that the lock is actually effective (in preventing changes - via REST API, classic UI, OfficeConnector and OfficeOnline). Please check whether this actually falls into Plone's responsibility and is covered by Plone tests that we should be able to rely on, or possibly some of ours in the meeting code.

<details><summary>The <b>move action</b> currently seems to be allowed (offered) on a locked document, but fails when attempting to perform it <i>(expand for traceback)</i></summary>
<p>

```
Traceback (most recent call last):
  File "/Users/lukasgraf/Plone/eggs/Zope2-2.13.28-py2.7.egg/ZPublisher/Publish.py", line 138, in publish
    request, bind=1)
  File "/Users/lukasgraf/Plone/eggs/Zope2-2.13.28-py2.7.egg/ZPublisher/mapply.py", line 77, in mapply
    if debug is not None: return debug(object,args,context)
  File "/Users/lukasgraf/Plone/eggs/Zope2-2.13.28-py2.7.egg/ZPublisher/Publish.py", line 48, in call_object
    result=apply(object,args) # Type s to step into published object.
  File "/Users/lukasgraf/Plone/eggs/plone.z3cform-0.8.1-py2.7.egg/plone/z3cform/layout.py", line 66, in __call__
    self.update()
  File "/Users/lukasgraf/Plone/eggs/plone.z3cform-0.8.1-py2.7.egg/plone/z3cform/layout.py", line 50, in update
    self.form_instance.update()
  File "/Users/lukasgraf/Plone/eggs/z3c.form-3.2.11-py2.7.egg/z3c/form/form.py", line 228, in update
    self.actions.execute()
  File "/Users/lukasgraf/Plone/eggs/plone.app.z3cform-0.7.8-py2.7.egg/plone/app/z3cform/csrf.py", line 21, in execute
    super(AuthenticatedButtonActions, self).execute()
  File "/Users/lukasgraf/Plone/eggs/z3c.form-3.2.11-py2.7.egg/z3c/form/action.py", line 98, in execute
    result = handler()
  File "/Users/lukasgraf/Plone/eggs/z3c.form-3.2.11-py2.7.egg/z3c/form/button.py", line 315, in __call__
    return handler(self.form, self.action)
  File "/Users/lukasgraf/Plone/eggs/z3c.form-3.2.11-py2.7.egg/z3c/form/button.py", line 170, in __call__
    return self.func(form, action)
  File "/Users/lukasgraf/Plone/buildouts/opengever/og-linkeddocs/opengever/dossier/move_items.py", line 159, in handle_submit
    failed_resource_locked_objects, )
  File "/Users/lukasgraf/Plone/buildouts/opengever/og-linkeddocs/opengever/dossier/move_items.py", line 231, in create_statusmessages
    mapping={'item': failed_objects[0]})
IndexError: list index out of range
``` 

</p>
</details>

File: opengever/locking/locales/de/LC_MESSAGES/opengever.locking.po [hunk 1]
```diff
@@ -33,3 +38,33 @@ msgstr "Dieses Dokument wurde als Kopie von ${document} eingereicht und kann nic
...
+
+#. Default: "Warning"
+#: ./opengever/locking/browser/templates/unlock_document_form.pt
+msgid "label_warning"
+msgstr "Warnung"
+
+#. Default: "Do you really want to unlock the document copied to a workspace?"
+#: ./opengever/locking/browser/unlock_forms.py
+msgid "msg_unlock_document_copied_to_workspace"
+msgstr "Das Dokument des Verknüpftem Teamraum kann dann immer zurückgeführt werden und als neue Version für dieses Dokument verwendet werden."
```
[the remark this answers is not in the document]
Comment (lukasgraf) [reply opengever/locking/locales/de/LC_MESSAGES/opengever.locking.po:55]: I've looked at this message in context, and there is indeed space for an explanatory message. I would suggest something like this:
`Änderungen, welche an diesem Dokument in GEVER vorgenommen werden, könnten durch ein Zurückführen des Dokuments aus Teamraum als neue Version überschrieben werden.`
But I could also see a much simpler message, and having a go at it in a second pass.

**Review by lukasgraf — APPROVED** [review]

**Comment by lukasgraf** [comment]
Changed my review to "Approved" because we already discussed the review feedback in detail, and there's nothing fundamental blocking a merge, IMHO. I would probably tend towards ship it _after_ the upcoming release.

**Comment by njohner** [comment]
@lukasgraf I've added tests for locked documents and fixed the behaviour for trashing and moving locked documents. Want to have another look?

**Comment by tinagerber** [comment]
@njohner changelog entry is in wrong release section

**Comment by lukasgraf** [comment]
@njohner Yep. If possible, I'd like to ship a hotfix before merging this one.

**Review by lukasgraf — APPROVED** [review]
LGTM! 👍 
Nice! I these tests and fixes actually cover quite a few paths that were already wonky before.

**Pull request merged**